File: webapp/lib/phases/setup_phase.rb
# frozen_string_literal: true

require_relative "../dungeon"
require_relative "../room"

# Setup phase (runs once at game start). The player makes the choices:
#   - they are dealt a starting hand (mulliganed until it holds a room)
#   - they are dealt two boss candidates and CHOOSE one (the other is discarded)
#   - they CHOOSE one room from hand to place beside their boss
#
# This phase exposes the discrete steps the game's decision loop drives; it does
# not pick for the player. Orchestration only.
class SetupPhase
  STARTING_ROOMS = 4
  STARTING_ABILITIES = 2
  BOSS_CANDIDATES = 2

  # Deal opening hands and draw each player's boss candidates. After this, every
  # player is awaiting a boss choice and a first-room choice.
  def self.deal(game)
    game.players.each do |player|
      deal_starting_hand(game, player)
      game.ability_deck.draw_many(STARTING_ABILITIES).each { |a| player.add_ability_to_hand(a) }
      game.set_boss_candidates(player, game.boss_deck.draw_many(BOSS_CANDIDATES))
    end
  end

  # Deal a starting hand, mulliganing (redrawing) until it contains at least one
  # basic room — only a basic room can be placed as the first room.
  def self.deal_starting_hand(game, player)
    loop do
      drawn = game.room_deck.draw_many(STARTING_ROOMS)
      if drawn.any? { |card| card.is_a?(Room) && !card.advanced? }
        drawn.each { |card| player.add_room_to_hand(card) }
        return
      end

      drawn.each { |card| game.room_deck.discard(card) } # mulligan
    end
  end

  # Apply a player's boss choice: keep the chosen card, discard the rest.
  def self.choose_boss(game, player, boss_id)
    candidates = game.boss_candidates(player)
    chosen = candidates.find { |boss| boss.id == boss_id }
    raise ArgumentError, "boss not among candidates: #{boss_id.inspect}" unless chosen

    (candidates - [chosen]).each { |boss| game.boss_deck.discard(boss) }
    player.dungeon = Dungeon.new(chosen)
    game.clear_boss_candidates(player)
  end

  # Apply a player's first-room choice (mandatory; must be an actual room).
  def self.place_first_room(_game, player, room_id)
    room = player.take_room_from_hand(room_id)
    raise ArgumentError, "room not in hand: #{room_id.inspect}" unless room
    raise ArgumentError, "first placement must be a room" unless room.is_a?(Room)

    player.dungeon.add_room_to_left(room)
  end
end

File: android/app/src/main/java/com/dungeonboss/game/phases/SetupPhase.kt
package com.dungeonboss.game.phases

import com.dungeonboss.game.Dungeon
import com.dungeonboss.game.Game
import com.dungeonboss.game.Player
import com.dungeonboss.model.Room

/**
 * Setup phase (runs once at game start). The player makes the choices: dealt a
 * starting hand (mulliganed until it holds a room), dealt two boss candidates
 * and CHOOSEs one, then CHOOSEs one room to place beside their boss. This phase
 * exposes the discrete steps the decision loop drives; it does not pick.
 */
object SetupPhase {
    const val STARTING_ROOMS = 4
    const val STARTING_ABILITIES = 2
    const val BOSS_CANDIDATES = 2

    /** Deal opening hands and draw each player's boss candidates. */
    fun deal(game: Game) {
        for (player in game.players) {
            dealStartingHand(game, player)
            game.abilityDeck.drawMany(STARTING_ABILITIES).forEach { player.addAbilityToHand(it) }
            game.setBossCandidates(player, game.bossDeck.drawMany(BOSS_CANDIDATES))
        }
    }

    /**
     * Deal a starting hand, mulliganing (redrawing) until it contains at least
     * one basic room — only a basic room can be placed as the first room (not an
     * upgrade or an advanced room).
     */
    fun dealStartingHand(game: Game, player: Player) {
        while (true) {
            val drawn = game.roomDeck.drawMany(STARTING_ROOMS)
            if (drawn.any { it is Room && !it.advanced }) {
                drawn.forEach { player.addRoomToHand(it) }
                return
            }
            drawn.forEach { game.roomDeck.discard(it) } // mulligan
        }
    }

    /** Apply a player's boss choice: keep the chosen card, discard the rest. */
    fun chooseBoss(game: Game, player: Player, bossId: String) {
        val candidates = game.bossCandidatesFor(player)
        val chosen = candidates.firstOrNull { it.id == bossId }
            ?: throw IllegalArgumentException("boss not among candidates: $bossId")

        candidates.filter { it !== chosen }.forEach { game.bossDeck.discard(it) }
        player.dungeon = Dungeon(chosen)
        game.clearBossCandidates(player)
    }

    /**
     * Apply a player's first-room choice (mandatory; must be an actual room),
     * placing it into the chosen slot (any of the 5).
     */
    fun placeFirstRoom(player: Player, roomId: String, slot: Int = 0) {
        val card = player.takeRoomFromHand(roomId)
            ?: throw IllegalArgumentException("room not in hand: $roomId")
        val room = card as? Room
            ?: throw IllegalArgumentException("first placement must be a room")
        player.dungeon!!.placeRoom(slot, room)
    }
}
